### PR TITLE
WIP: Add bitbanging I2C driver.

### PR DESCRIPTION
**Pull request by pfalcon**
This is not intended to be merged as is, rather to be an illustration for further New HW API discussion.

Also, is it good directory name to be put to? (IMHO, organization of stuff isn't ideal, but I don't have immediate plan how to improve it).

### CODE REVIEW

File: drivers/i2c-bitbang/i2c.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+from machine import Pin
+
+
+class I2C:
+
+    def __init__(self, sda, scl):
+        sda.init(Pin.IN)
+        scl.init(Pin.IN)
+        sda(0)
+        scl(0)
```
Comment (dpgeorge) [inline drivers/i2c-bitbang/i2c.py]: Why did you not set sda, scl to self.sda, self.scl?  Why use the global variables?

File: drivers/i2c-bitbang/i2c.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        self.delay();
+        bit = sda();
+        self.scl_low();
+        return bit
+
+    def writebit(self, val):
+        if val:
+            self.sda_high();
+        else:
+            self.sda_low();
```
Comment (dpgeorge) [inline drivers/i2c-bitbang/i2c.py]: This is an example of why I think it's better to just do self.sda(val).

**Comment by dpgeorge** [comment]
I think making sda_high() etc methods is the wrong approach.  Instead just call the pin with 0 or 1.  Eg:
```python
def readbit(self):
    self.sda(1)
    self.scl(0)
    self.delay()
    bit = self.sda()
    self.scl(0)
    return bit
```
Then you can remove the 4 sda/scl/high/low functions.  It's still just as configurable (if not more) because you can define your sda/scl "pin" objects to do whatever you like when passed 0/1.  They could be in OPEN_DRAIN mode and it'd work.  scl could be in OUT mode and it'd work just the same.  Or you can wrap a real pin in adaptor methods to do other things.  The job of the I2C driver is to call pin(0) or pin(1), not to reconfigure the mode of the pin.

File: drivers/i2c-bitbang/i2c.py (hunk 1 above)
Comment (pfalcon) [inline drivers/i2c-bitbang/i2c.py]: LOL. Indeed, funny glitch. That was straightforward translated of metaprogramming C++ code, and those args were template params. Of course, that needs to be fixed.

File: drivers/i2c-bitbang/i2c.py (hunk 2 above)
Comment (pfalcon) [inline drivers/i2c-bitbang/i2c.py]: So, this class uses 2 dozens of high/low calls, you pointed at 2 which could be "optimized" into 1 .value() call. Doesn't really sound too convincing ;-).

**Comment by pfalcon** [comment]
> Then you can remove the 4 sda/scl/high/low functions.

These are primitives of controlling pins, each of these translates to a single machine instruction. Anything else is built on top of these operations, and this code shows this (also to a beginner who studies this stuff).

> They could be in OPEN_DRAIN mode and it'd work.

There's no OPEN_DRAIN mode. It's optional, many MCUs don't implement it. The implementation in STM32 is particularly anomalous, most MCUs have explicit direction bit and don't allow to read pin with OUT direction. So, you just try force this anomalous scheme as default for MicroPython ;-). No wonder people scratch their heads, looking in datasheets and saying "open drain? what open drain? there's no open drain in my mcu!"

> Or you can wrap a real pin in adaptor methods to do other things.

Yes, that's definitely should be supported, but if course not practical for performance. This code tries to both show how it works and be actually usable.

> The job of the I2C driver is to call pin(0) or pin(1), not to reconfigure the mode of the pin.

Not exactly either. There's difference between "pin" and "signal". Pin can have values 0 and 1, whereas signal can be asserted or deasserted. How signal states map to pin states - depends, this example this example shows that it maps to pin direction.


Again, what you're trying to do is to hide good deal of this stuff behind the covers of HW API, which can only have effects: 1) not letting people to study how it really works at barebone level (no matter if you think that open-drain should be such barebone level - matter of fact that not every MCU has it); 2) confuse people (again, if there's no open drain for particular MCU, there's no it, even if you think it's vendor conspiracy and there's actually OD in every MCU); 3) performance implications - there's no need to test for modes, values, etc. For basic stuff, the code knows large part of the state, there's no need to duplicate it somewhere and hurt performance.

File: drivers/i2c-bitbang/i2c.py (hunk 2 above)
Comment (dpgeorge) [inline drivers/i2c-bitbang/i2c.py]: My point is that all calls to sda/scl/high/low are removed and replaced with just .value().

**Comment by dpgeorge** [comment]
> These are primitives of controlling pins, each of these translates to a single machine instruction.

You translated this from PTL, which is why it's not matching the Python HW API.  These methods you defined do not (and will never) translate to single machine instructions in uPy.  The reason for `__call__()` in the HW API is so that you can efficiently provide such logic level switching.

> There's no OPEN_DRAIN mode. It's optional, many MCUs don't implement it

There is.  It's defined in the HW API.  Ports should implement it.

> "open drain? what open drain? there's no open drain in my mcu!"

Who says this?  Pyboard has OD and that's where uPy runs.

> The job of the I2C driver is to call pin(0) or pin(1), not to reconfigure the mode of the pin.
> 
> Not exactly either. There's difference between "pin" and "signal". Pin can have values 0 and 1, whereas signal can be asserted or deasserted. How signal states map to pin states - depends, this example this example shows that it maps to pin direction.

The job of the I2C driver is to provide correct timing and logic level shifting.  The job of the "pin" that you pass in is to provide an abstraction of the electrical states that correspond to a "0" and "1".  What you are doing in the above code is mixing the timing & logic level shifting with the electrical states.

> Again, what you're trying to do is to hide good deal of this stuff behind the covers of HW API

Yes, the aim of a good API is to provide an abstraction and hide stuff that you don't want to worry about.  The HW API provides IN/OUT/OD as abstractions.

> 1) not letting people to study how it really works at barebone level (no matter if you think that open-drain should be such barebone level - matter of fact that not every MCU has it); 2) confuse people

On the contrary, having OD teaches people something: that there is more to life than just IN/OUT.  It lets them understand better the way busses like I2C operate, and why they need pull ups.

> 3) performance implications

Making OD first class in the HW API is exactly so we can have better performance.  See the 1-wire driver.

File: drivers/i2c-bitbang/i2c.py (hunk 2 above)
Comment (pfalcon) [inline drivers/i2c-bitbang/i2c.py]: I don't understand such a point. I can understand if you say they **can** be replaced. That's true. And what I say is that they can be NOT replaced. And I don't see any contradiction here, or why you think your way is infallibly better.

**Comment by pfalcon** [comment]
> You translated this from PTL, which is why it's not matching the Python HW API.

But PTL matches actual hardware. And the whole point of this patch was to show that it's possible to implement I2C without calling for OD pins, so it's a bold claim that "it's not matching" ;-).

> These methods you defined do not (and will never) translate to single machine instructions in uPy.

That's pretty bleak picture you draw. Better argument would be that if uPy had enough machinery to translate .high() to a single instruction, then it almost certainly could .translate .value(1) either. And countering that is still easy: for .high(), simple inlining is enough, for .value(1) - not. Getting closer to reality, my code can be translate to inline assembler, where each .high(), etc. is a single asm instruction - for majority of MCUs. One you propose - not directly (would need to do constant propagation and dead code removal to arrive at single instruction).

> Who says this? Pyboard has OD and that's where uPy runs.

So, we've talked about pyboard JW API so far? Let's stop and talk about arbitrary MCUs and beyond them. Good reference which should be always kept in mind as a checklist: https://www.kernel.org/doc/Documentation/gpio/gpio.txt (count works like "vary" and "some").

> The job of the I2C driver is to provide correct timing and logic level shifting. The job of the "pin" that you pass in is to provide an abstraction of the electrical states that correspond to a "0" and "1". What you are doing in the above code is mixing the timing & logic level shifting with the electrical states.

I don't know why you take that as undeniable truth. Generally, the most simplest concepts are usually defined as combination of different notions (try to define what 5 is - it's superset of all individual definitions possible). So, you say that bus defines just timing and layered on top of pins which are binary switchable. What I say that bus controls it all, and is able to push pin into any state - 0, 1, or Z, not just in 1 of 2 preconfigured states. I hint that my scheme is more flexible than yours, and thus forms better basis (and that shows - you don't have OPEN_SOURCE defined in you API, while I can do it easily). But then what I ask is that it can be implemented efficiently, not to remove yours, whereas you push that mine don't even make sense to exist.

> Yes, the aim of a good API is to provide an abstraction and hide stuff that you don't want to worry about.

People want to worry, as you see.

> On the contrary, having OD teaches people something:

I've learnt a lot. Now if you expect me to forget what I knew before that, then it's not how it works.

> It lets them understand better the way busses like I2C operate, and why they need pull ups.

Looking at that from position of high impedance doesn't teach people false or even less.

> Making OD first class in the HW API is exactly so we can have better performance. See the 1-wire driver.

Translated to C, direction-based scheme will have same or better performance (while saving need to emulate OD).

**Comment by pfalcon** [comment]
Patch was updated to use member variables.

**Comment by dpgeorge** [comment]
Closing as I don't see that this will be merged.

**Pull request closed without merging**